elastic/beats
# Implement backoff for Kafka output

## What does this PR do?

Implements the equal-jitter backoff strategy on the Kafka output.

Note that we cannot simply use `outputs.WithBackoff` on the Kafka output client because the latter's [`Publish()` method](https://github.com/elastic/beats/blob/cc6c4e33db3fdcf2bd7d962532c14fc080d5371b/libbeat/outputs/kafka/client.go#L129-L158) never returns an `error`. The [wrapper `Publish()` method](https://github.com/elastic/beats/blob/95626b8f1690344312c0831ab2bdcbccffe4d089/libbeat/outputs/backoff.go#L59-L66) implemented by the backoff client depends on the underlying output client returning an `error` so it can block (i.e. backoff for) a certain duration before returning.

So instead we use `sarama`'s [`Producer.Retry.BackoffFunc`](https://github.com/Shopify/sarama/blob/1c3eb669b9e55f80d2861cef8dfd921f88d868cd/config.go#L230) while reusing the same backoff API that is used by other libbeat network outputs, viz. Elasticsearch, Logstash, and Redis.

## Why is it important?

It will allow users to configure backoff options (`backoff.init` and `backoff.max`) for the Kafka output similar to what they can do for the Elasticsearch, Logstash, and Redis outputs.

## Checklist

- [ ] ~I have added tests that prove my fix is effective or that my feature works~ An automated test is quite cumbersome as it requires the orchestration of several processes. So for now I've listed out the manual test steps below.

## How to test this PR locally

To test this PR, one needs to cause the partition leader node to go away, which should trigger the backoff. One also needs to subsequently bring the partition leader node back online to ensure that the backoff duration has been reset.

1. Start a 3-node Kafka cluster. Follow the instructions through step 6 on https://kafka.apache.org/quickstart but skip steps 3, 4, and 5.

2. Configure `filebeat.yml` (could be any Beat, but Filebeat is easy for testing) like so:
   ```
   filebeat.inputs:
   - type: log
     enabled: true
     paths:
       - /tmp/logs/*.log

   output.kafka:
     enabled: true
     hosts: [ "localhost:9092", "localhost:9093", "localhost:9094" ]
     version: 2.1
     topic: foobar
     backoff:
       init: 3s
       max: 30s
   ```

3. Start writing log lines to the test log file.
   ```
   mkdir /tmp/logs
   while true; do echo "$RANDOM new log entry" >> foo.log; sleep 0.1; done
   ```

4. Start Filebeat.
   ```
   ./filebeat -c filebeat.test.yml -e
   ```

5. Check that the `foobar` topic has been created in Kafka.
   ```
   bin/kafka-topics.sh --describe --bootstrap-server localhost:9092 --topic segerbeger
   ```

   This topic is expected to have only 1 partition, partition `0`. Note this partition's leader node ID.

6. Now kill the Kafka node corresponding to the partition leader's node ID.

7. Re-run the command in step 5 and verify that that partition `0` has no leader.

8. Check the Filebeat log. You should now see entries like this:
```
2020-04-22T07:06:35.852-0700    INFO    [kafka] kafka/config.go:288     backing off for 4.715506403s (init: 3s, max: 30s)
2020-04-22T07:06:42.094-0700    INFO    [kafka] kafka/config.go:288     backing off for 7.428653822s (init: 3s, max: 30s)
2020-04-22T07:06:51.053-0700    INFO    [kafka] kafka/config.go:288     backing off for 14.693773032s (init: 3s, max: 30s)
2020-04-22T07:07:07.272-0700    INFO    [kafka] kafka/config.go:288     backing off for 16.638039354s (init: 3s, max: 30s)
```
9. Re-start the node you killed in step 6.

10. Check the Filebeat log again. The "backing off for" entries should stop appearing in the log.

11. Repeat steps 6. through 10. Make sure that the backoff duration is reset to a low number (close to the `backoff.init` value) again.

12. Stop all 6 processes (ZK, 3x Kafka nodes, log generator, FB) to end the test.

## Related issues

- Resolves elastic/beats#16777
